test/test_preprocess.py

```diff
@@ -325,50 +328,172 @@ def test_preprocess_pipeline(tmpdir):
+                    if joiner_end:
+                        tok_replace[-1] += joiner_marker
+                    assert (tokens[0][pos:pos+len(tok_replace)] == tok_replace)
+
+            def change_align_side(al_idx, pos, num_to_del, tok_replace):
+                new_al_idx = al_idx
+                len_tok_replace = len(tok_replace) if tok_replace else 0
+                if al_idx >= pos:
+                    if al_idx < pos+num_to_del:
+                        # insertion
+                        if len_tok_replace:
+                            new_al_idx = pos
+                        # deletion
+                        else:
+                            return None
+                    else:
+                        # shift
+                        new_al_idx = al_idx - num_to_del + len_tok_replace
+                return new_al_idx
+
+
+            src_tokens = tu.src_tok.tokens if tu.src_tok else None
+            tgt_tokens = tu.tgt_tok.tokens if tu.tgt_tok else None
+
+            src_len = len(src_tokens[0]) if src_tokens else 0
+            tgt_len = len(tgt_tokens[0]) if tgt_tokens else 0
+
+            if src_len and tgt_len:
+                alignment_before = deepcopy(tu.alignment[0])
+
+                joiner_marker = "￭"
+
+                next(self._rand_repl_gen)
+                src_replace, tgt_replace = self._rand_repl_gen.send((src_len, tgt_len))
+
+                src_pos, src_num_to_del, src_tok_replace = src_replace
+                tgt_pos, tgt_num_to_del, tgt_tok_replace = tgt_replace
+
+                src_joiner_start, src_joiner_end = joiner_side(src_tokens, src_pos, src_num_to_del)
+                tgt_joiner_start, tgt_joiner_end = joiner_side(tgt_tokens, tgt_pos, tgt_num_to_del)
+
+                tu.replace_tokens(src_replace, tgt_replace)
+
+                checks_side(src_tokens, src_len, src_pos, src_num_to_del, src_tok_replace, src_joiner_start, src_joiner_end)
+                checks_side(tgt_tokens, tgt_len, tgt_pos, tgt_num_to_del, tgt_tok_replace, tgt_joiner_start, tgt_joiner_end)
+
+                # Check alignment
+                alignment_after = tu.alignment[0]
+
+                new_align = set()
+                for al_src, al_tgt in alignment_before :
+                    new_al_src = change_align_side(al_src, src_pos, src_num_to_del, src_tok_replace)
+                    if new_al_src is None:
+                        continue
+                    new_al_tgt = change_align_side(al_tgt, tgt_pos, tgt_num_to_del, tgt_tok_replace)
+                    if new_al_tgt is None:
+                        continue
+                    new_align.add((new_al_src, new_al_tgt))
+
+                assert new_align == alignment_after
+            return [tu]
+
+    config_replace = deepcopy(config_base)
+
+    config_replace['preprocess'].append(
+        { "op": "repl" }
+    )
+
+    preprocessor = TrainingProcessor(config_replace, "", str(tmpdir))
```

Comment:
Is a Preprocessor needed to test the token replacement feature? Or is it possible to build custom TUs and test the operator directly?

Reply:
The preprocessor is not directly needed to test the function itself, but it allows sampling sentences of various lengths and applying all the necessary steps (tokenization, alignment).